### ISSUE TITLE
Problem with recordings in Windows host

### ISSUE THREAD

**Issue by bonigarcia** [aerokube/selenoid#329]
Hello,

I use the following command to start selenoid as Docker container withrecording capabilities:

`docker run --rm --name selenoid -p 4444:4444 -v //var/run/docker.sock:/var/run/docker.sock -v /c/Users/boni/selenoid/config/:/etc/selenoid/:ro -v /c/Users/boni/selenoid/video/:/opt/selenoid/video/ -e OVERRIDE_VIDEO_OUTPUT_DIR=/c/Users/boni/selenoid/video/ aerokube/selenoid:latest-release`

(of course the folder `selenoid/config` and `selenoid/video` exists in my Windows home, and the proper `browser.json` file is contained in `selenoid/config`)

Selenoid starts fine and wait for incoming requests in port 4444. Then, I run the following JUnit 4 test case:

```
public class RemoteTest {

    private WebDriver driver;

    @Before
    public void setupTest() throws MalformedURLException {
        DesiredCapabilities capability = DesiredCapabilities.chrome();
        capability.setCapability("enableVideo", true);
        driver = new RemoteWebDriver(
                new URL("http://192.168.99.100:4444/wd/hub"), capability);
    }

    @After
    public void teardown() {
        if (driver != null) {
            driver.quit();
        }
    }

    @Test
    public void test() {
        driver.get("https://en.wikipedia.org/wiki/Main_Page");
    }

}
```

Everything works fine except the recording. I can see the following error trace in the selenoid container logs:

`2018/01/09 10:25:03 [0] [VIDEO_ERROR] [Failed to rename /opt/selenoid/video/selenoid279b9066911947c322f41491f79c838c.mp4 to /opt/selenoid/video/b15b8602ecbf02912c98bc257a08a7e7.mp4: rename /opt/selenoid/video/selenoid279b9066911947c322f41491f79c838c.mp4 /opt/selenoid/video/b15b8602ecbf02912c98bc257a08a7e7.mp4: text file busy]`

And in the host folder `selenoid/video` only the temporal recording appears (`selenoid279b9066911947c322f41491f79c838c.mp4` in this example), but the expected recording file, which should be `<session-id>.mp4` is not (as shown in the error trace).

![image](https://user-images.githubusercontent.com/5413013/34716781-2bffed9a-f531-11e7-97d9-720b26db9eff.png)

A bug maybe?

**Comment by vania-pooh** [comment]
@bonigarcia could you give more details on your Selenoid version as well as Windows version?

**Comment by bonigarcia** [comment]
I'm using Windows 10 Home edition. My docker engine has been installed through the [Docker Toolbox](https://docs.docker.com/toolbox/toolbox_install_windows/) (not [Docker for Windows](https://www.docker.com/docker-windows), which requires the Pro edition of Windows).

Regarding the Selenoid version, as you can see in the command, is `aerokube/selenoid:latest-release`.

**Comment by vania-pooh** [comment]
Hmm... One more question. Before this error line there should be two lines like this:
```
2018/01/09 10:25:03 [0] [STOPPING_VIDEO_CONTAINER] [<id>]
2018/01/09 10:25:03 [0] [STOPPED_VIDEO_CONTAINER] [<id>]
```
Or probably the second line could be:
```
2018/01/09 10:25:03 [0] [FAILED_TO_STOP_VIDEO_CONTAINER] [<id>] [<some-message>]
```
Could you copy these lines, too?

**Comment by bonigarcia** [comment]
Sure, the complete log is as follows :

```
2018/01/09 10:24:42 Loading configuration files...
2018/01/09 10:24:42 Loaded configuration from [/etc/selenoid/browsers.json]
2018/01/09 10:24:42 Using default containers log configuration because of: read error: open config/container-logs.json: no such file or directory
2018/01/09 10:24:42 Timezone: Local
2018/01/09 10:24:42 Video Dir: /opt/selenoid/video
2018/01/09 10:24:42 Listening on :4444
2018/01/09 10:25:00 [NEW_REQUEST]
2018/01/09 10:25:00 [NEW_REQUEST_ACCEPTED]
2018/01/09 10:25:00 [0] [LOCATING_SERVICE] [chrome-]
2018/01/09 10:25:00 Using default version: 63.0
2018/01/09 10:25:00 [0] [USING_DOCKER] [chrome-63.0]
2018/01/09 10:25:00 [0] [CREATING_CONTAINER] [selenoid/vnc:chrome_63.0]
2018/01/09 10:25:00 [0] [STARTING_CONTAINER] [selenoid/vnc:chrome_63.0] [51d88e3f514d529355b7b74b5677c8f2ba0c8da03b7340d4c93a302dd11855c4]
2018/01/09 10:25:00 [0] [CONTAINER_STARTED] [selenoid/vnc:chrome_63.0] [51d88e3f514d529355b7b74b5677c8f2ba0c8da03b7340d4c93a302dd11855c4] [167.524977ms]
2018/01/09 10:25:00 [0] [CREATING_VIDEO_CONTAINER] [selenoid/video-recorder]
2018/01/09 10:25:00 [0] [STARTING_VIDEO_CONTAINER] [selenoid/video-recorder] [808eb3e7679065f3d288b9380ab6ad23916d57b814a23fcdc20bc78d81043c75]
2018/01/09 10:25:01 [0] [VIDEO_CONTAINER_STARTED] [selenoid/video-recorder] [808eb3e7679065f3d288b9380ab6ad23916d57b814a23fcdc20bc78d81043c75] [230.973116ms]
2018/01/09 10:25:01 [0] [SERVICE_STARTED] [selenoid/vnc:chrome_63.0] [51d88e3f514d529355b7b74b5677c8f2ba0c8da03b7340d4c93a302dd11855c4] [669.035µs]
2018/01/09 10:25:01 [0] [PROXY_TO] [selenoid/vnc:chrome_63.0] [51d88e3f514d529355b7b74b5677c8f2ba0c8da03b7340d4c93a302dd11855c4] [http://172.17.0.3:4444]
2018/01/09 10:25:01 [0] [SESSION_ATTEMPTED] [unknown] [http://172.17.0.3:4444] [1]
2018/01/09 10:25:02 [0] [SESSION_CREATED] [unknown] [b15b8602ecbf02912c98bc257a08a7e7] [http://172.17.0.3:4444] [1] [1.416237073s]
2018/01/09 10:25:03 [SESSION_DELETED] [b15b8602ecbf02912c98bc257a08a7e7]
2018/01/09 10:25:03 [0] [STOPPING_VIDEO_CONTAINER] [808eb3e7679065f3d288b9380ab6ad23916d57b814a23fcdc20bc78d81043c75]
2018/01/09 10:25:03 [0] [STOPPED_VIDEO_CONTAINER] [808eb3e7679065f3d288b9380ab6ad23916d57b814a23fcdc20bc78d81043c75]
2018/01/09 10:25:03 [0] [REMOVING_CONTAINER] [51d88e3f514d529355b7b74b5677c8f2ba0c8da03b7340d4c93a302dd11855c4]
2018/01/09 10:25:03 [0] [CONTAINER_REMOVED] [51d88e3f514d529355b7b74b5677c8f2ba0c8da03b7340d4c93a302dd11855c4]
2018/01/09 10:25:03 [0] [VIDEO_ERROR] [Failed to rename /opt/selenoid/video/selenoid279b9066911947c322f41491f79c838c.mp4 to /opt/selenoid/video/b15b8602ecbf02912c98bc257a08a7e7.mp4: rename /opt/selenoid/video/selenoid279b9066911947c322f41491f79c838c.mp4 /opt/selenoid/video/b15b8602ecbf02912c98bc257a08a7e7.mp4: text file busy]
```

**Comment by vania-pooh** [comment]
Very interesting. Video container is stopped as expected. Any stuck containers in `docker ps` output?

**Comment by bonigarcia** [comment]
No, after the test only the container named `selenoid` is up and running.

**Comment by vania-pooh** [comment]
Ok, let me then try to reproduce myself. Need to search for Windows somewhere. :)

**Comment by baranoff-ivan** [comment]
Got same issue on Windows 7 + Docker Toolbox

**Comment by vania-pooh** [comment]
@baranoff-ivan I'm still failing to find any computer with Windows to check myself. All my notebooks have no virtualization support. :(

**Comment by khilghard** [comment]
I've got the same bug, but this time on a Linux system.

2018/03/20 21:06:32 [9] [CONTAINER_STARTED] [selenoid/vnc:firefox_58.0] [932ff58f22777a284a5030381505ff3eed3ca7a83eb06e2880e095bf31218e43] [0.38s]
2018/03/20 21:06:32 [9] [CREATING_VIDEO_CONTAINER] [selenoid/video-recorder]
2018/03/20 21:06:33 [9] [STARTING_VIDEO_CONTAINER] [selenoid/video-recorder] [2c1e13675cfc7a30c0665d541e2f334e8e1a515b1901dd2016fedc98d3167b2e]
2018/03/20 21:06:33 [9] [VIDEO_CONTAINER_STARTED] [selenoid/video-recorder] [2c1e13675cfc7a30c0665d541e2f334e8e1a515b1901dd2016fedc98d3167b2e] [0.36s]
2018/03/20 21:06:33 [9] [SERVICE_STARTED] [selenoid/vnc:firefox_58.0] [932ff58f22777a284a5030381505ff3eed3ca7a83eb06e2880e095bf31218e43] [0.00s]
2018/03/20 21:06:33 [9] [PROXY_TO] [932ff58f22777a284a5030381505ff3eed3ca7a83eb06e2880e095bf31218e43] [http://172.17.0.6:4444/wd/hub]
2018/03/20 21:06:33 [9] [SESSION_ATTEMPTED] [http://172.17.0.6:4444/wd/hub] [1]
2018/03/20 21:06:35 [9] [SESSION_CREATED] [8019c4bc-9bec-4a8b-aa40-68d1db0cffd2] [1] [2.69s]
2018/03/20 21:06:37 [16] [SESSION_DELETED] [8019c4bc-9bec-4a8b-aa40-68d1db0cffd2]
2018/03/20 21:06:37 [9] [STOPPING_VIDEO_CONTAINER] [2c1e13675cfc7a30c0665d541e2f334e8e1a515b1901dd2016fedc98d3167b2e]
2018/03/20 21:06:37 [9] [STOPPED_VIDEO_CONTAINER] [2c1e13675cfc7a30c0665d541e2f334e8e1a515b1901dd2016fedc98d3167b2e]
2018/03/20 21:06:37 [9] [REMOVING_CONTAINER] [932ff58f22777a284a5030381505ff3eed3ca7a83eb06e2880e095bf31218e43]
2018/03/20 21:06:37 [9] [CONTAINER_REMOVED] [932ff58f22777a284a5030381505ff3eed3ca7a83eb06e2880e095bf31218e43]
2018/03/20 21:06:37 [9] [VIDEO_ERROR] [Failed to rename /video/selenoid607667f7e1c7923779e35506b040300d.mp4 to /video/8019c4bc-9bec-4a8b-aa40-68d1db0cffd2.mp4: rename /video/selenoid607667f7e1c7923779e35506b040300d.mp4 /video/8019c4bc-9bec-4a8b-aa40-68d1db0cffd2.mp4: no such file or directory]

**Comment by khilghard** [comment]
In the selenoid container, I get:
drwxr-xr-x    2 15210197 15209600    4.0K Mar 20 21:06 .
drwxr-xr-x    3 root     root        4.0K Mar 20 21:05 ..
-rw-r--r--    1 root     root      613.9K Mar 20 21:06 selenoid607667f7e1c7923779e35506b040300d.mp4
-rw-r--r--    1 root     root      305.7K Mar 20 21:06 selenoid6a305e75138c4c5ce32508dc7e46832f.mp4
-rw-r--r--    1 root     root      269.7K Mar 20 21:02 selenoid7d99a2a51fc843b294d2703e36b53917.mp4
-rw-r--r--    1 root     root      586.2K Mar 20 21:03 selenoide17f0da4236ef77ef6f11fc15fe569b8.mp4
/opt/selenoid/video # 

On my host os, I get:
-rw-r--r-- 1 root    root         614K Mar 20 15:06 selenoid607667f7e1c7923779e35506b040300d.mp4
-rw-r--r-- 1 root    root         306K Mar 20 15:06 selenoid6a305e75138c4c5ce32508dc7e46832f.mp4
-rw-r--r-- 1 root    root         270K Mar 20 15:02 selenoid7d99a2a51fc843b294d2703e36b53917.mp4
-rw-r--r-- 1 root    root         587K Mar 20 15:03 selenoide17f0da4236ef77ef6f11fc15fe569b8.mp4

**Comment by vania-pooh** [comment]
@khilghard you forgot on of the following:
1) Set `OVERRIDE_VIDEO_OUTPUT_DIR` variable to host machine video dir
2) Set `-video-output-dir` to `/opt/selenoid/video` and mount host machine dir to this mount point

**Comment by khilghard** [comment]
I didn't realize that the directory name had to be exactly /opt/selenoid/video. We started using selenium and switched to selenoid but kept the selenium directory name. Now that I have switched the directory name to selenoid, it works fine. That would be a good thing to put in the documentation. Here was the command I used to make my selenoid docker work for us:

docker run --name selenoid -d -p 4445:4444 -v /var/run/docker.sock:/var/run/docker.sock -v /opt/selenoid/:/etc/selenoid/:ro -v /opt/selenoid/video/:/opt/selenoid/video/ -e OVERRIDE_VIDEO_OUTPUT_DIR=/opt/selenoid/video/ aerokube/selenoid:latest-release -limit 50 -conf /etc/selenoid/browsers.json -video-output-dir /opt/selenoid/video

All of my configuration files were placed in the /opt/selenoid directory.

@vania-pooh  Thanks for the assist!

**Comment by vania-pooh** [comment]
@khilghard http://aerokube.com/selenoid/latest/#_video_recording could you please take a look and comment - whether something is not clear?

**Comment by khilghard** [comment]
I read the documentation as allowing for a custom output directory which I tried, but got the above failure. Then you mentioned using /opt/selenoid/video which I switched to using both configuration variables. All that I changed was the location to both OVERRIDE_VIDEO_OUTPUT_DIR and -video-output-dir and that made things work. The directory was correct but the '/' char was missing between the directory and filename. If I use what appears to be a hard coded setting, everything worked great. If I wanted to use a custom directory that didn't seem to work.

It seems as though a custom directory can be used, but that was difficult to get to work. Perhaps a simple note similar to the above for me about troubleshooting video not showing up with the error I got would be sufficient.

Perhaps something like this:
If an error similar to the following error occurs (docker logs -f [docker hash]):
2018/03/20 21:06:37 [9] [VIDEO_ERROR] [Failed to rename /video/selenoid607667f7e1c7923779e35506b040300d.mp4 to /video/8019c4bc-9bec-4a8b-aa40-68d1db0cffd2.mp4: rename /video/selenoid607667f7e1c7923779e35506b040300d.mp4 /video/8019c4bc-9bec-4a8b-aa40-68d1db0cffd2.mp4: no such file or directory]

It may be that the following settings need to be checked, adjusted or added (if one of the options was not set):
1. Set OVERRIDE_VIDEO_OUTPUT_DIR variable to host machine video dir
2. Set -video-output-dir to /opt/selenoid/video and mount host machine dir to this mount point

It wasn't immediately clear to me how the mapping was done, but after your comments, I was able to figure it out.

**Comment by vania-pooh** [comment]
@khilghard will be added to FAQ.

**Comment by bonigarcia** [comment]
@vania-pooh Just FYI, I have just tried the latest version of Selenoid (i.e. 1.6.0), and the problem with the recordings in Windows is still there :(

**Comment by vania-pooh** [comment]
@bonigarcia that's logical. The issue is not closed yet. We applied our effort to deliver file download feature as well as other fixes. All stuff included to milestones is available on the Issues page in this repository.

**Comment by vania-pooh** [comment]
This should have been fixed in 2311596eb9cdd0907a1baaf3ad05632b0bb6857e. Now we wait for video container to stop before renaming the file.
